### ISSUE TITLE
I find cntk new version is so slow for rnn, and I don't konw the reason. I use old version is so fast.

### ISSUE THREAD

**Issue by whaozl** [Microsoft/CNTK#2187]
**this minibatch-size=128**
**This model is LSTM. but one minibatch need 2 seconds.**

```
Selected GPU[0] TITAN Xp as the process wide default device.
2017-08-04 11:45:38,635 ============[/data/anjos/anjos.data/cntk-pos/train]============
2017-08-04 11:45:44,773 ============sentences=200000 minibatch-size=128
2017-08-04 11:45:46,682 ============[/data/anjos/anjos.data/cntk-pos/test]============
2017-08-04 11:45:48,593 ============sentences=52388 minibatch-size=128
2017-08-04 11:46:13,022 Epoch[   0] Minibatch[  10]: [Training]time= 0:00:22.375134, loss=  3.0121,metric=  0.8055
2017-08-04 11:46:34,524 Epoch[   0] Minibatch[  20]: [Training]time= 0:00:21.502004, loss=  2.9661,metric=  0.8385
2017-08-04 11:46:55,552 Epoch[   0] Minibatch[  30]: [Training]time= 0:00:21.027393, loss=  2.8747,metric=  0.8146
2017-08-04 11:47:15,645 Epoch[   0] Minibatch[  40]: [Training]time= 0:00:20.092026, loss=  2.8591,metric=  0.7947
2017-08-04 11:47:35,928 Epoch[   0] Minibatch[  50]: [Training]time= 0:00:20.283260, loss=  2.8586,metric=  0.8417
2017-08-04 11:47:56,719 Epoch[   0] Minibatch[  60]: [Training]time= 0:00:20.790460, loss=  2.6338,metric=  0.7777
2017-08-04 11:48:18,395 Epoch[   0] Minibatch[  70]: [Training]time= 0:00:21.675715, loss=  2.3150,metric=  0.6610
2017-08-04 11:48:39,625 Epoch[   0] Minibatch[  80]: [Training]time= 0:00:21.229870, loss=  2.2929,metric=  0.6378
2017-08-04 11:49:01,851 Epoch[   0] Minibatch[  90]: [Training]time= 0:00:22.225212, loss=  2.0622,metric=  0.6071
2017-08-04 11:49:22,471 Epoch[   0] Minibatch[ 100]: [Training]time= 0:00:20.619831, loss=  1.8843,metric=  0.5209
2017-08-04 11:49:42,169 Epoch[   0] Minibatch[ 110]: [Training]time= 0:00:19.697690, loss=  1.8157,metric=  0.4915
2017-08-04 11:50:02,055 Epoch[   0] Minibatch[ 120]: [Training]time= 0:00:19.886112, loss=  1.6115,metric=  0.4251
```

**Comment by KeDengMS** [comment]
@whaozl Please use [profiler ](https://docs.microsoft.com/en-us/cognitive-toolkit/BrainScript-and-Python-Performance-Profiler) to check what's slowed down comparing to old version. If possible, please share a repro. Thanks!

**Comment by boeddeker** [comment]
@KeDengMS is it possible to extend the wiki under the link to contain an example, what the user can expect from the profiler? The last time I used the profiler, it didn't gave me the expected output, so I wasted some time to figure this out.

**Comment by whaozl** [comment]
@KeDengMS @boeddeker  I use the logging to calculate the time of each batch, the model is as follows:
···
def unbi_model_seq2seq(num_embed,num_lstm_layer,num_hidden,label_vocab_len):
    with cntk.layers.default_options(init=xavier()):
        model = Sequential([
            Label('input'),
            Embedding(num_embed, name='embed'),
            Label('embeded_input'),
            Stabilizer(name='stabilizer1'),
            For(range(num_lstm_layer), lambda i: [
                Recurrence(LSTM(num_hidden,name='lstm_%d'%i))
            ]),
            Stabilizer(name='stabilizer2'),
            Dropout(0.5),
            Label('hidden_representation'),
            Dense(label_vocab_len, name='label_out'),
        ])
    return model
···
This input and label of model are:
```
label_vocab_len=98
data_vocab_len=40000
z=unbi_model_seq2seq(200,2,200,label_vocab_len)
input_var = sequence.input_variable(shape=data_vocab_len, sequence_axis=Axis('inputAxis'))
label_var = sequence.input_variable(shape=label_vocab_len, sequence_axis=Axis('inputAxis'))
pe = cross_entropy_with_softmax(z, label_var)
errs = classification_error(z, label_var)
```

I use mxnet, in same condition, one minibatch(128) is so fast, need 0.01~0.05s

**Comment by KeDengMS** [comment]
@whaozl Are you running on GPU or CPU? What's the length of sequence? I tried your script with random sequence length with max of 100, and 128 sequences. It runs on TitanX with 0.134s per minibatch for 2.1, and 0.130 in 2.0.

Here's the output of profiler:
_Minibatch Iteration      :       133.983 ms         0.812 ms       133.486 ms       137.229 ms               19      2545.674 ms
__Forward + Backward      :       133.216 ms         0.811 ms       132.718 ms       136.460 ms               19      2531.105 ms
__Weight Update           :         0.743 ms         0.004 ms         0.740 ms         0.759 ms               19        14.123 ms

Here _Minibatch Iteration shows the total cost in training one minibatch, which includes _Forward + Backward and _Weight Update (update the model using learner).

Some suggestions on speed:
1. I assume your input_var is sparse, given the high dimension of data_vocab_len. Switch to sparse would be much faster than dense.
2. If you are training with GPU, try use [optimized_rnnstack](https://www.cntk.ai/pythondocs/cntk.ops.html?highlight=optimized_rnnstack#cntk.ops.optimized_rnnstack) instead of Recurrence(LSTM()), which leverages CuDNN. It usually is 3x faster than layers.LSTM. You can use [optimized_rnnstack_converter](https://www.cntk.ai/pythondocs/cntk.misc.optimized_rnnstack_converter.html?highlight=optimized_rnnstack#module-cntk.misc.optimized_rnnstack_converter) if you need to run the model on CPU

@boeddeker That's a good suggestion, we'll work on a manual/tutorial on profiler usage.

**Comment by whaozl** [comment]
@KeDengMS I found that  it takes 2s about a batch whenthe number of label categories is 99 of Chinese POS tagging, the speed as your almost when  the number of label categories is 10. I run your test benchmark code, the speed is basically the same. Is this speed related to the sample?

**Comment by KeDengMS** [comment]
@whaozl Please check the profiler output and see if _Forward + Backward changes in your scenario. I think the slowdown may come from data input.

**Comment by whaozl** [comment]
@KeDengMS Thank you very much, Can you provide a example about input and label are sparse, I only found input is sparse in cntk, I tried the input and label sparse, but had the error.

**Comment by KeDengMS** [comment]
We may have unimplemented loss functions for sparse label. If you are using CrossEntropyWithSoftmax you should be OK to use sparse label.

**Comment by whaozl** [comment]
Thank you very much!
